### ISSUE TITLE
Question about Sampling of 'q_z_xy'

### ISSUE THREAD

**Issue by SqySimmon** [StanfordASL/Trajectron-plus-plus#15]
Thanks for your nice work.
I am a little confused about training a VAE model without sampling latent variable z mentioned in paper: 
> Notably, the Gumbel-Softmax reparameterization is not used to backpropagate through the Categorical latent variable z because it is not sampled during training time.

The inputs of decoder are every one-hot samples of Categorical latent variable, which are invariant during training.
https://github.com/StanfordASL/Trajectron-plus-plus/blob/b99e07dcabba06aab14135a79ac19e77d2baa314/trajectron/model/components/discrete_latent.py#L31-L35
https://github.com/StanfordASL/Trajectron-plus-plus/blob/b99e07dcabba06aab14135a79ac19e77d2baa314/trajectron/model/components/discrete_latent.py#L95-L97

**Comment by BorisIvanovic** [comment]
This is a great question @SqySimmon , thank you for asking. It's a finer point in the paper so I'm glad someone picked up on it!

There are two ways that loss gradients backpropagate to the encoder part of the model (with the standard CVAE ELBO loss function):
1. Through the `KL(q || p)` term
2. Through reparameterized z samples in the `E[ log p(y | x, z) ]` term

Doing the full z enumeration scheme above, the only avenue of backpropagation that is affected is 2 (avenue 1 is only based off the logits of p and q).

Since we have a tractable discrete latent space (|Z| = 25), we can directly compute the expectation over `log p(y | x, z)`. It ends up being `\sum_z q(z | x, y) log p(y | x, z)`. The first term in this sum is just the q probabilities (which we have in the q logits) and the second term is the log probability of the ground truth under our generated distribution (which is easy to obtain since our model outputs the mus and sigmas for each z value). 

Since gradients backpropagate through this loss term, they backpropagate through the q logits (which are the important values during training if you were sampling z from q). Since the encoder directly generates the q logits then we have the gradients making their way back to the encoder, as desired.

**Comment by JoeHEZHAO** [comment]
Dear Author

Thanks for your sharing.  I still have some questions on this topic.

As you mentioned, function **sample_q()** would be invariant during training. However, it seems that function **sample_p()** is also invariant if flag **full_dist** or **all_z_sep** are True as in:

https://github.com/StanfordASL/Trajectron-plus-plus/blob/f8a11fd90a1a4cb64c70ece0d4cf4a92dfc4e7f7/trajectron/model/components/discrete_latent.py#L37-L62

Since the eval_loss() and predict() functions of mgcvae.py clearly set the flag as **full_dist**, 

https://github.com/StanfordASL/Trajectron-plus-plus/blob/f8a11fd90a1a4cb64c70ece0d4cf4a92dfc4e7f7/trajectron/model/mgcvae.py#L1071
`        def eval_loss()...
z = self.latent.sample_p(1, mode, full_dist=True)
`

https://github.com/StanfordASL/Trajectron-plus-plus/blob/f8a11fd90a1a4cb64c70ece0d4cf4a92dfc4e7f7/trajectron/model/mgcvae.py#L1083-L1096

**Does this mean **z** will always be invariant ?? ** 

Thanks for your time !!

**Comment by JoeHEZHAO** [comment]
Oh, sorry now I understand.

You set the flag to False in train.py when evaluating and visualizing.

**Comment by mvignaga** [comment]
Dear Ivan, 
I have a question about your first answer in this issue, where you say that we backpropagate through the q logits. I'm looking for it in your code but I could only find the sample_q function returning the all_one_hot vector, which is then given as input to the GRU in p_y_xz.
Could you please explain where the q logits are involved in the training loss?
